### ISSUE TITLE
Mondrian forests based optimizer

### ISSUE THREAD

**Issue by glouppe** [scikit-optimize/scikit-optimize#356]
@MechCoder did a terrific job reimplementing Mondrian Forests on top of the Scikit-Learn codebase, at https://github.com/scikit-optimize/mondrian_forest

Now that this is working, I think it is time to implement the corresponding minimizer. 

Questions:
* Shall this live as an option in `forest_minimize`? or rather as its own `mondrian_minimize`?
* To reduce dependencies, shall we backport the  `mondrian_forest` codebase à la joblib in scikit-learn? or rather add it as new dependency? 

CC: @betatim @balajiln

**Comment by betatim** [comment]
> * To reduce dependencies, shall we backport the  `mondrian_forest` codebase à la joblib in scikit-learn? or rather add it as new dependency? 

not sure what you mean with backport a la joblib :confused: 

The best option would be to have this merged into sklearn: most exposure, we get help maintaining it, no new dependency.

If we think getting it merged into scikit-learn isn't going to happen then let's keep it as a separate package instead of adding the code to scikit-optimize directly. I would say if mondrian_forest doesn't have dependencies of itself that are new, we should add a new dependency for skopt on mondrian_forest.

How much of the tree code in this repo is a copy & paste from the sklearn tree code? I didn't look too carefully, at first glance it seems that at least it has a common starting point. In which case we should try and get it merged into sklearn to reduce the effort of mondrian_forest profiting from the updates that others make to sklearn's tree code. Or is there a way to include `pyx` files from sklearn when building mondiran_forest somehow?

(Basically I am in favour of the option that leaves us/@mechcoder with the least amount of code to maintain :grin:)

> * Shall this live as an option in `forest_minimize`? or rather as its own `mondrian_minimize`?

I was wondering if we could better organise the various `XXX_minimize` functions and reduce the number of parameters each of them take. Feels like we already have too many parameters ... which points towards adding more `foo_minimize` functions which have less parameters (because they are fixed for the user).

---

Maybe you mean we add the code to skopt directly under `skopt.externals` like sklearn did for joblib?

**Comment by glouppe** [comment]
> Maybe you mean we add the code to skopt directly under skopt.externals like sklearn did for joblib?

Yes, this is the option I had in mind.

**Comment by MechCoder** [comment]
See https://github.com/scikit-optimize/scikit-optimize/issues/359 with regards to having a lot of parameters.

I would be in favour of adding it as an optional dependency with the `MondrianForestRegressor` as a default base estimator and falling back onto `ExtraTreesRegressor` if it is not installed.

**Comment by MechCoder** [comment]
The advantage being to use the regressors in places outside of hyperparameter optimisation.
